worker/src/main/java/gov/cms/ab2d/worker/adapter/bluebutton/PatientClaimsProcessorImpl.java

```diff
@@ -0,0 +1,138 @@
+import java.util.concurrent.Future;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.stream.Collectors;
+
+@Slf4j
+@Component
+@RequiredArgsConstructor
+public class PatientClaimsProcessorImpl implements PatientClaimsProcessor {
+
+    @Autowired
+    private BFDClient bfdClient;
+
+    @Autowired
+    private FhirContext fhirContext;
+
+    @Autowired
+    private FileService fileService;
+
+    @Async("bfd-client")
+    public Future<Integer> process(String patientId, ReentrantLock lock, Path outputFile, Path errorFile) {
+        int errorCount = 0;
+        int resourceCount = 0;
+
+        try {
+            var resources = getEobBundleResources(patientId);
+
+            var jsonParser = fhirContext.newJsonParser();
+
+            var byteArrayOutputStream = new ByteArrayOutputStream();
+            for (var resource : resources) {
+                ++resourceCount;
+                try {
+                    final String payload = jsonParser.encodeResourceToString(resource) + System.lineSeparator();
+                    byteArrayOutputStream.write(payload.getBytes(StandardCharsets.UTF_8));
+                } catch (Exception e) {
+                    ++errorCount;
+                    handleException(errorFile, e, lock);
+                }
+            }
+
+            appendToFile(outputFile, byteArrayOutputStream, lock);
+        } catch (Exception e) {
+            ++errorCount;
+            try {
+                handleException(errorFile, e, lock);
+            } catch (IOException e1) {
+                //should not happen - original exception will be thrown
+                log.error("error during exception handling to write error record");
+            }
+            throw new RuntimeException(e);
+        }
+
+        log.info("finished writing [{}] resources", resourceCount);
+
+        if (errorCount > 0) {
+            log.warn("[{}] error records. Should create an error row in JobOutput table", errorCount);
+        }
+
+        return new AsyncResult<>(errorCount);
+    }
+
+    private void handleException(Path errorFile, Exception e, ReentrantLock lock) throws IOException {
+        var errMsg = ExceptionUtils.getRootCauseMessage(e);
+        var operationOutcome = FHIRUtil.getErrorOutcome(errMsg);
+
+        var jsonParser = fhirContext.newJsonParser();
+        var payload = jsonParser.encodeResourceToString(operationOutcome) + System.lineSeparator();
+
+        var byteArrayOutputStream = new ByteArrayOutputStream();
+        byteArrayOutputStream.write(payload.getBytes(StandardCharsets.UTF_8));
+        appendToFile(errorFile, byteArrayOutputStream, lock);
+    }
+
+    /**
+     * uses re-entrant lock to lock the shared file resource
+     * @param outputFile
+     * @param byteArrayOutputStream
+     * @throws IOException
+     */
+    private void appendToFile(Path outputFile, ByteArrayOutputStream byteArrayOutputStream, ReentrantLock lock) throws IOException {
+        lock.lock();
```

Comment:
We should use `tryLock(long timeout, TimeUnit unit)` instead. Imagine a different thread has acquired the lock, but got stuck writing JSON to the output file due to I/O issues (EFS is a network file system). All of the other threads that reach this line will "freeze" trying to acquire the lock. 

Let's make the timeout configurable via `application.properties` and set it to the default of 30 seconds. 

If `tryLock` returns `false`, let's consider this a non-recoverable error, throw an exception and thus terminate any further processing of the `Future`.

Reply:
done